### PR TITLE
Add serverless beats tests

### PR DESCRIPTION
**Pull request by fearful-symmetry**
## What does this PR do?

This is (hopefully) the final stage of getting serverless testing working. This Adds 17 serverless tests, as well as a mage role to run the tests. Assuming you have a beats binary built, you can run `mage integration:testBeatServerless metricbeat`

This does not include any buildkite steps to run the tests; I'm not totally sure how we want to schedule them, and if that should be part of this PR.

## Why is it important?

We need serverless tests.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [x] I have added an integration test or an E2E test

### CODE REVIEW

File: testing/integration/beats_serverless_test.go
Comment (pchila) [inline testing/integration/beats_serverless_test.go:5]: Having just the `integration` build tag will include this file while trying to launch integration tests for elastic agent. Maybe we want to specify an additional tag like:

```suggestion
//go:build integration
//go:build serverless
```
Ofc we need to add the tag in magefile target

File: testing/integration/beats_serverless_test.go
Comment (pchila) [inline testing/integration/beats_serverless_test.go:134]: Shouldn't this conditioned to the url scheme being https (which is 99.9% of the time is)? It's just that setting fixed `:443` to a URL where we don't even check the scheme feels wrong...
Furthermore, since we spent the time parsing the URL couldn't we set `parsedKibana.Host` field and then call `parsedKibanaHost.String()` ?

File: testing/integration/beats_serverless_test.go
Comment (pchila) [inline testing/integration/beats_serverless_test.go:148]: maybe we can use a `template` instead of `fmt.Sprintf` ? Having positional arguments substituted in the config kinda makes me nervous

File: testing/integration/beats_serverless_test.go
Comment (pchila) [inline testing/integration/beats_serverless_test.go:168]: We are only doing it once in the whole test, let's keep it simple and allocate one extra string ;)
```suggestion
	docs, err := estools.GetLatestDocumentMatchingQuery(ctx, runner.requirementsInfo.ESClient, map[string]interface{}{
		"match": map[string]interface{}{
			"host.test-id": runner.testUuid,
		},
	}, "*"+runner.testbeatName+"*")
```

File: testing/integration/beats_serverless_test.go
Comment (fearful-symmetry) [reply testing/integration/beats_serverless_test.go:140]: You would think, but the beats defaults to `5601` if you don't specify the port, hence why we need to add it manually.

File: testing/integration/beats_serverless_test.go
Comment (fearful-symmetry) [reply testing/integration/beats_serverless_test.go:5]: Yah, considering that this PR is technically temporary until we get buildkite setup in the beats repo, I didn't want to get to clever with anything that would require modifications to the agent code, since eventually this'll become a proper integration test in the agent repo.

**Comment by cmacknz** [comment]
Given that this is finding actual bugs in serverless Beats, my preference to unblock this is to  run these tests on a scheduled https://github.com/elastic/elastic-agent/issues/3470

Otherwise we will be blocked waiting for the Beats repository's migration to Buildkite which is probably 1+ month away.

**Comment by fearful-symmetry** [comment]
@cmacknz agreed. Is adding some scheduled run in buildkite something that can be done in this PR, or is there some other global config?

**Comment by cmacknz** [comment]
You should be able to do it in this PR

File: .buildkite/pipeline.yml
Comment (cmacknz) [inline .buildkite/pipeline.yml:166]: Where is the cron schedule controlled? Can you link to it here?

**Comment by cmacknz** [comment]
Do you have a link to the Beats tests passing on Buildkite, since they aren't running as part of the PR automation?

**Comment by fearful-symmetry** [comment]
@cmacknz here: https://buildkite.com/elastic/elastic-agent/builds/4490#018b87a1-0854-4559-b675-20e3033f15e4

File: .buildkite/pipeline.yml
Comment (fearful-symmetry) [reply .buildkite/pipeline.yml:166]: So, I was going to add it once the PR was merged, but the control panel for it is here: https://buildkite.com/elastic/elastic-agent/settings/schedules

File: testing/integration/beats_serverless_test.go
Comment (cmacknz) [reply testing/integration/beats_serverless_test.go:5]: Changing the build tag wouldn't just require you to make a dedicated mage target for running these I think. That is probably a good thing.

How are these getting excluded from `mage integration:test` runs today?

File: .buildkite/pipeline.yml
Comment (cmacknz) [reply .buildkite/pipeline.yml:191]: Add a link to that as a comment right above this so future us can find it easily

File: .buildkite/scripts/steps/beats_tests.sh
Comment (cmacknz) [inline .buildkite/scripts/steps/beats_tests.sh:70]: We can skip testing auditbeat since our team doesn't maintain it. None of the code we are testing is really Beat specific regardless.

File: .buildkite/scripts/steps/beats_tests.sh
Comment (cmacknz) [inline .buildkite/scripts/steps/beats_tests.sh:39]: Why do we need to clone and build Beats?

Why don't we just download the latest SNAPSHOT image using the artifacts API?

**Comment by cmacknz** [comment]
Your sample test run doesn't contain any Beat logs. If these tests fail how do we debug them?

There are no test artifacts generated either https://buildkite.com/elastic/elastic-agent/builds/4490#018b87a1-0854-4559-b675-20e3033f15e4

So there is no test report generated to make the failures easier to see.

It also looks like it is running every other test. Is that the intent for the scheduled job? I would have it only run the serverless Beats test so it isn't going to fail for any other random reason.

**Review by pchila — COMMENTED** [review]
A couple of comments left about the pipeline step script and the mage target.
Overall code is ok, I am not too fond of the mixing of beats serverless and elastic agent integration tests being guarded only by a condition on the TestSuite and a mage target that would allow to run those tests against a stateful stack (not sure that we have a usecase for that).

File: .buildkite/scripts/common.sh
Comment (fearful-symmetry) [reply .buildkite/scripts/common.sh:56]: so, because we define a `mage` bash function that encases the actual mage executable and we only want to check if the actual binary is installed, `type -p` seems to be the only command I found that behaves in the way the script initially expected. The previous version would always return true.

File: .buildkite/scripts/steps/beats_tests.sh
Comment (fearful-symmetry) [reply .buildkite/scripts/steps/beats_tests.sh:62]: ah, good point!

File: magefile.go
Comment (fearful-symmetry) [reply magefile.go:1587]: It's occasionally useful for testing, usually for cases if we aren't sure if something is just broken, or only broken under serverless. Has happened a lot in the past.

**Comment by fearful-symmetry** [comment]
@pchila not sure what you mean? The `if runner.testbeatName == "elastic-agent" ` line was just the simplest way to constrain the tests, considering this entire PR is meant to be reverted once buildkite is available in the beats repo.

**Review by pchila — APPROVED** [review]

**Pull request merged**
